RFCs/0033-graph-query.md

```diff
@@ -0,0 +1,491 @@
+within PartiQL as a `MATCH` expression form.
+
+Consequently, our goal here is to describe how the two semantics are combined,
+specifically addressing:
+- Role of PartiQL environments while evaluating pattern matching.
+- Evaluation of PartiQL expressions within graph `WHERE` clauses.
+- And, crucially, how the result of pattern match computation is
+  represented as a value within the PartiQL data model.
+
+
+For evaluating a  graph matching expression
+>    `(` _graph_ `MATCH` _pattern_ `)`
+
+this RFC  proposes the following.
+
+- As any PartiQL expression, the graph matching expression
+  is evaluated within a regular PartiQL environment (𝝆<sub>0</sub>, 𝝆) that maps variables
+  (defined outside this expression) to values.
+
+- The _graph_ subexpression is evaluated within (𝝆<sub>0</sub>, 𝝆) and must result
+  in a graph value.
+
+- Given the graph value, graph matching computation for _pattern_ proceeds according
+  to the GPML semantics, with the following modifications:
+
+  - Variables defined in _pattern_ shadow variables bound in (𝝆<sub>0</sub>, 𝝆).
+
+  - Whenever a `WHERE e` clause within the pattern needs to be evaluated,
+    the clause's expression `e` is evaluated within (𝝆<sub>0</sub>, 𝝆)
+    though regular PartiQL evaluation, except that any pattern-defined
+    variable `x` is not looked up in (𝝆<sub>0</sub>, 𝝆).
+    Instead, its value, as a graph element, is provided by the GPML
+    pattern matching procedure.  
+    The subsequent processing of the graph-element value depends on the
+    context of `x`:
+    - If `x` is used within a graph-specific construct (such as a predicate
+      like `IS DIRECTED` or a path-aggregating function), the graph-element
+      value is passed to this construct directly.
+    - Otherwise, `x` is used as a regular PartiQL value.
+      - If `x` is bound to a node or an edge, the payload value of the latter
+        is used.
+      - If `x` is bound to a path, it is an error.
+
+- Upon completion of the GPML pattern matching computation, the resulting
+  bag of matching graph fragments is converted into a bag of structs
+  by converting each fragment into a struct as follows.  
+  Given a fragment of the form
+  >  _p<sub>1</sub>_ *,* ...*,* _p<sub>n</sub>_
+
+  where _p<sub>i</sub>_ are variable-annotated paths,
+  the resulting struct gets key/value attribute pairs where keys are variable names
+  while the attribute pair for variable `x` is determined is follows.
+
+  - If the `x` is a singleton variable that annotates a node or an edge,
+    the value of its attribute is the payload value at the node or edge.
+  - If `x` is a group variable, the value of its attribute is
+    the list of payloads at the elements (nodes or edges) annotated with `x`,
+    taken in the order of their appearance in the fragment's path.
+
+  - If `x` is a path variable, it does not produce an attribute in the struct,
+    according to this RFC.
```

Comment:
Why do path variables not bind to an attribute?  Are they purely internal to the `MATCH`?

Reply:
Well, pattern variables are "internal" to a MATCH in the same sense as others (node and edge variables) -- in the sense that, for the outside of MATCH all their "graphiness" is stripped away and only payloads play role.  
But for that, for a path variable specifically, there is no clear choice for that "payload residual":  

- Overall, this probably should be some list drawn from the elements (nodes and edges) of the path.
- But should it contain only the elements of the path named by node/edge variables or include anonymous elements as well?  
- Should it be an independent list, or be a super-structure that incorporates the above bindings for nodes/elements?
  - In the former case, do we want to represent the fact that some parts of the path's list are also appearing at other variables?  How to do that reliably?
  - In the latter case, this would be a hierarchical structure (due to nested paths), making access to nodes and edges not as uniform as in the proposed (lines 417-421).
- Or should the structure bound to the path contain only the anonymous elements, because the named ones are represented already?

While it could be possible to cobble together a "reasonable" solution based on some choices from the above, the choices would be essentially arbitrary and likely a misfit with what would be eventually coming in SQL/PGQ. 

This was my reason for not proposing anything in this regard in this RFC.  But perhaps a reference from here to the "To Be Resolved" section would be helpful (plus the section itself).

Reply:
There is now a section on this in Unresolved Questions, which might also be a tad more coherent then the above comment.